src/common/snippets/src/lowered/pass/identify_buffers.cpp

```diff
@@ -116,75 +120,86 @@ std::vector<bool> IdentifyBuffers::create_adjacency_matrix(const LinearIR& linea
+    const auto output_count = loop_end->get_output_num();
+
+    const auto& ptr_increments = loop_end->get_ptr_increments();
+    const auto& finalization_offsets = loop_end->get_finalization_offsets();
+    const auto& data_sizes = loop_end->get_element_type_sizes();
+
+    BufferMap buffer_neighbours;
+    for (size_t i = 0; i < input_count; ++i) {
+        const auto& parent_output = loop_end_expr->get_input_port_connector(i)->get_source().get_expr();
+        if (ov::is_type<op::Buffer>(parent_output->get_node())) {
+            if (buffer_neighbours.count(parent_output) > 0) {
+                OPENVINO_ASSERT(buffer_neighbours[parent_output].ptr_increment == ptr_increments[i] &&
+                                buffer_neighbours[parent_output].finalization_offset == finalization_offsets[i],
+                                "Invalid data pointer shifts: If Buffer has several consumers, this consumers must have the same shifts or zero");
+                continue;
             }
+            buffer_neighbours[parent_output] = { data_sizes[i], ptr_increments[i], finalization_offsets[i] };
         }
-
-        for (auto buffer_it = buffer_neighbours.begin(); buffer_it != buffer_neighbours.end(); ++buffer_it) {
-            for (auto neighbour_it = std::next(buffer_it); neighbour_it != buffer_neighbours.end(); ++neighbour_it) {
-                update_adj_matrix(*buffer_it, *neighbour_it, pool, adj);
+    }
+    for (size_t i = input_count; i < input_count + output_count; ++i) {
+        // The consumers of the corresponding Store ops
+        const auto consumer_inputs = loop_end_expr->get_input_port_connector(i)->get_consumers();
+        size_t buffer_count = 0;
+        size_t loop_count = 0;
+        for (const auto& consumer_input : consumer_inputs) {
+            const auto& child_expr = consumer_input.get_expr();
+            if (ov::is_type<op::Buffer>(child_expr->get_node())) {
+                buffer_neighbours[child_expr] = { data_sizes[i], ptr_increments[i], finalization_offsets[i] };
+                buffer_count++;
+            } else if (ov::is_type<op::LoopEnd>(child_expr->get_node())) {
+                loop_count++;
             }
         }
+        if (buffer_count > 0) {
+            OPENVINO_ASSERT((buffer_count == 1) && (buffer_count + loop_count == consumer_inputs.size()),
+                            "Loop output must have not more than 1 Buffer");
+        }
     }
+    return buffer_neighbours;
+}
 
-    return adj;
+IdentifyBuffers::BufferMap IdentifyBuffers::get_buffer_loop_inside(const LinearIR::constExprIt& loop_end_it) {
+    // Buffers which are connected to the current Loop but without ptr shifts and Buffers which are inside this Loop - must be adjacent because
+    // after each Loop iteration GPR will be shifted using ptr increment of Buffer outside. But Buffers inside have the same GPR - it means that
+    // Buffers inside will work with shifted memory.
```

Comment:
It looks like in bounds of `get_buffer_loop_inside` we work only with buffers inside the Loop. However, this comment contains information about Buffers which are outside the current Loop too, right? If yes, maybe it's better to move this information to the place where we work with buffers outside the loop?

Reply:
Good idea! Moved, thanks!